backend/store/etcd/watchers.go

```diff
@@ -81,86 +68,78 @@ func (s *Store) GetCheckConfigWatcher(ctx context.Context) <-chan store.WatchEve
 
 // GetEntityWatcher returns a channel that emits WatchEventEntity structs notifying
 // the caller that an Entity was updated. If the watcher runs into a terminal error
-// or the context passed is cancelled, then the channel will be closed. The caller must
-// restart the watcher, if needed.
+// or the context passed is cancelled, then the channel will be closed.
+// The watcher does its best to recover from errors.
 func (s *Store) GetEntityWatcher(ctx context.Context) <-chan store.WatchEventEntity {
 	ch := make(chan store.WatchEventEntity, 1)
-	wc := s.client.Watch(ctx, entityKeyBuilder.Build(""), clientv3.WithPrefix(), clientv3.WithCreatedNotify())
+	key := entityKeyBuilder.WithContext(ctx).Build()
+	w := Watch(ctx, s.client, key, true)
+
 	go func() {
 		defer close(ch)
+		for response := range w.Result() {
+			if response.Type == store.WatchUnknown {
+				logger.Error("unknown etcd watch type: ", response.Type)
+				continue
+			}
 
-		for resp := range wc {
-			for _, event := range resp.Events {
-				action := GetWatcherAction(event)
-				if action == store.WatchUnknown {
-					logger.Errorf("unknown etcd watch action: %s", event.Type.String())
-					continue
-				}
-
-				var entity corev2.Entity
-				if action != store.WatchDelete {
-					if err := json.Unmarshal(event.Kv.Value, &entity); err != nil {
-						logger.WithError(err).Error("error unmarshaling watch event")
-						continue
-					}
-				} else {
-					// Fill in the name and namespace of the deleted entity
-					parts := strings.Split(string(event.Kv.Key), "/")
-					if len(parts) > 1 {
-						entity.Name = parts[len(parts)-1]
-						entity.Namespace = parts[len(parts)-2]
-					}
-				}
+			var entity corev2.Entity
 
-				watchEvent := store.WatchEventEntity{
-					Action: action,
-					Entity: &entity,
+			if response.Type == store.WatchDelete {
+				meta := store.ParseResourceKey(response.Key)
+				entity.Namespace = meta.Namespace
+				entity.Name = meta.ResourceName
+			} else {
+				if err := json.Unmarshal(response.Object, &entity); err != nil {
+					logger.WithField("key", response.Key).WithError(err).Error("unable to unmarshal entity from key")
+					continue
 				}
+			}
 
-				ch <- watchEvent
+			ch <- store.WatchEventEntity{
+				Action: response.Type,
+				Entity: &entity,
 			}
 		}
 	}()
+
 	return ch
 }
 
 // GetTessenConfigWatcher returns a channel that emits WatchEventTessenConfig structs notifying
 // the caller that a TessenConfig was updated. If the watcher runs into a terminal error
 // or the context passed is cancelled, then the channel will be closed. The caller must
```

Comment:
I wish you could do multi-line suggestions

```suggestion
// or the context passed is cancelled, then the channel will be closed.
```

Reply:
Thanks, I'll look into it